### PR TITLE
Virtual file fixes

### PR DESCRIPTION
**Pull request by mpickering**

### CODE REVIEW

File: src/Language/Haskell/LSP/VFS.hs [hunk 1]
```diff
@@ -62,35 +67,49 @@ data VirtualFile =
...
     , _text    :: Rope
-    , _tmp_file :: Maybe FilePath
     } deriving (Show)
 
-type VFS = Map.Map J.NormalizedUri VirtualFile
+type VFSMap = Map.Map J.NormalizedUri VirtualFile
+
+data VFS = VFS { vfsMap :: Map.Map J.NormalizedUri VirtualFile
+               , vfsTempDir :: FilePath -- ^ This is where all the temporary files will be written to
+               } deriving Show
```
Comment (alanz) [inline src/Language/Haskell/LSP/VFS.hs]: I think we should move the `revMap` in here too.

**Comment by alanz** [comment]
Apart from moving the `revMap`. LGTM. We could possibly do that as a separate PR.

File: src/Language/Haskell/LSP/Core.hs [hunk 2]
```diff
@@ -893,7 +916,6 @@ serverCapabilities clientCaps o h =
           J.CompletionOptions
             (Just $ isJust $ completionResolveHandler h)
             (map singleton <$> completionTriggerCharacters o)
-            (map singleton <$> completionAllCommitCharacters o)
```
Comment (mpickering) [inline src/Language/Haskell/LSP/Core.hs]: Did I do something wrong to have to change these lines?

File: src/Language/Haskell/LSP/VFS.hs (hunk 1 above)
Comment (mpickering) [reply src/Language/Haskell/LSP/VFS.hs]: Done

**Comment by mpickering** [comment]
@alanz I incorporated your changes into this PR (without the MVar)

**Comment by mpickering** [comment]
I somehow messed up the history quite badly by doing a merge. Let me know if you want the history to be cleaned up somehow.

**Comment by alanz** [comment]
LGTM

**Pull request merged**